File: app/channels/staging_area_channel.rb
class StagingAreaChannel < ApplicationCable::Channel
  # Called when the consumer has successfully become a subscriber to
  # this channel.
  def subscribed
    stream_from room_broadcast_identifier
    stream_from user_broadcast_identifier
  end

  # Called once the consumer has cut its cable connection.
  # Any cleanup needed when channel is unsubscribed.
  def unsubscribed
    # TODO
  end

  # NOTE: This method is also called from a user model callback.
  # @param :room (Room)
  def self.get_broadcast_data(room)
    data = {
      slug: room.slug,
      users: []
    }
    room.users.each do |user|
      data.dig(:users) << user
    end
    data
  end

  # @param :name (String) -> name to assign to current user.
  def set_name(data)
    current_user.name = data['name']
    # Once a user's name is set, they move from staging to the game room.
    if current_user.save
      room = current_user.room
      if room.users.count == 1 || room.drawer_id.nil?
        room.drawer_id = current_user.id
        room.save!
      end
      ActionCable.server.broadcast(
        room_broadcast_identifier,
        {
          context: 'refresh_players',
          payload: StagingAreaChannel.get_broadcast_data(room)
        }
      )
      ActionCable.server.broadcast(
        user_broadcast_identifier,
        { context: 'usher_to_game_room', slug: room.slug }
      )
    end
  end

private
  # pub/sub link for room.
  def room_broadcast_identifier(room_slug = params[:slug])
    "staging_area_#{room_slug}"
  end

  # pub/sub link for user.
  def user_broadcast_identifier(room_slug = params[:slug], user_id = current_user.id)
    "staging_area_#{room_slug}_#{user_id}"
  end
end

File: app/javascript/channels/staging_area_channel.js
import consumer from "channels/consumer"

(() => {
  const split = window.location.pathname.split('/');
  if (split[1] === 'staging_areas' && typeof split[2] === 'string') {
    const slug = split[2];
    const channel = consumer.subscriptions.create({ channel: 'StagingAreaChannel', slug: slug }, {
      // Called when the subscription is ready for use on the server
      connected() {
        console.log('staging_area - connected');
      },

      // Called when the subscription has been terminated by the server
      disconnected() {
        console.log('staging_area - disconnected');
      },

      // Called when there's incoming data on the websocket for this channel
      received(data) {
        switch (data.context) {
          case 'refresh_players':
            this.refreshPlayers(data.payload);
            break;
          case 'usher_to_game_room':
            this.redirectToGameRoom(data);
            break;
        }
      },

      refreshPlayers(data) {
        const range = document.createRange();
        let players = [];
        document.querySelector('h3').innerText = `Players in room '${data.slug}' (${data.users.length})`
        if (data.users.length > 0) {
          document.querySelector('#room-information').classList.remove('hidden');
          data.users.forEach((user) => {
            let playerCardHTML = `
              <div class="c-card animate-pop-in">
                <p class="title">${user.name}</p>
              </div>
            `;
            const playerCard = range.createContextualFragment(playerCardHTML);
            players.push(playerCard);
          });
          document.querySelector('.player-staging-container').replaceChildren(...players);
        } else {
          document.querySelector('#room-information').classList.add('hidden');
        }
      },

      redirectToGameRoom(data) {
        window.location.href = `/rooms/${data.slug}`;
      }
    });

    const nameInput = document.querySelector('#name_input');
    const submitButton = document.querySelector('form > button');

    const names = () => {
      return Array.from(
        document.querySelectorAll('.c-card > .title')
      ).map((node) => { return node.innerText; });
    }

    const sanitizeInput = (input) => {
      if (input) {
        return input.replace(/<(.|\n)*?>/g, '').trim();
      }
      return "";
    };

    const shake = () => {
      const form = document.querySelector('form');

      form.classList.add('invalid-input');
      setTimeout(() => {
        form.classList.remove('invalid-input');
      }, 500);
    }

    const setNameRequest = (name) => {
      if (names().includes(name) || name.length === 0) {
        if (name.length === 0) {
          document.querySelector('p.error').innerText = `Please enter a name.`;
        } else {
          document.querySelector('p.error').innerText = `
            '${name}' is already taken, please enter another.
          `;
        }
        shake();
        return;
      }
      // Calls 'StagingAreaChannel#set_name(data)' on the server.
      channel.perform('set_name', { name: name });
    };

    submitButton.addEventListener('click', (event) => {
      event.preventDefault();
      setNameRequest(sanitizeInput(nameInput.value));
    });
    nameInput.addEventListener('keydown', (event) => {
      if (event.code === 'Enter') {
        event.preventDefault();
        setNameRequest(sanitizeInput(event.target.value));
      }
    });
  }
})();
